campaignd: Blacklist standard reserved DOS Device Names

shikadilord opened the pull request:
Windows XP and later* define DOS device names for several system devices that will be handled as if they had directory entries everywhere. Their names may not be used on the left side of a file name if a dot is present, otherwise the read or write operation in progress (if applicable) is redirected to the device in question, with unexpected results (e.g. writing to `CON` writes to a console that may or may not be
attached, writing to `COM1` will send data through the first enumerated serial port, and so on).

Thus, we really don't want add-on authors to either intentionally or accidentally use them on non-Windows platforms and cause Windows-based clients to do unexpected things.

Emitting special messages for the UI in this case seems unnecessary and it'll be a waste of space most of the time since these names should crop up extremely rarely. Better just tell people what's up on a case-by-case basis if they decide to post in the forums about it.

\* (Yes, all Windows versions, really, although the DOS-based versions could have an arbitrary number of additional devices.)

jyrkive commented on src/addon/validation.cpp:
```diff
@@ -94,6 +104,11 @@ bool addon_filename_legal(const std::string& name)
 	   name.size() > 255) {
 		return false;
 	} else {
+		const std::string stem = boost::algorithm::to_upper_copy(name.substr(0, name.find('.')), std::locale::classic());
```
Please use [`filesystem::base_name()`](https://github.com/wesnoth/wesnoth/blob/4f225f56e0acd6b059dc5cbba20e85c6a905714c/src/filesystem.hpp#L190) to extract the filename instead.

shikadilord replied:
That's not going to give the expected results. For example, `con.test.test` is handled by Windows as if referring to `CON`, but `filesystem::basename()` is going to return `con.test` instead and therefore not match the list.

jyrkive replied:
I see. Please add a comment to mention that, then.

GregoryLundberg commented on src/addon/validation.cpp:
```diff
@@ -30,6 +30,16 @@ namespace {
 
+	// Reserved DOS device names on Windows XP and later.
+	const std::set<std::string> dos_device_names = {
+		"NUL", "CON", "AUX", "PRN",
```
CLOCK$
COM5 ... COM9
LPT5 ... LPT9

shikadilord replied:
CLOCK$ is not exposed to Windows applications. Is it possible for more than 4 serial or parallel ports to be enumerated?

GregoryLundberg replied:
Dunno about this millennium, but onceuponatime I had an 8-port serial card in a DOS box so it used to be so.

Frankly, all this seems way too close to the edge of the Rabbit Hole, and I'd suggest stopping before you eat any more of the Tea Cakes. There are times when it's better to ask the user which foot than to try to stop them shooting.

shikadilord replied:
You make it sound like this is about protecting Windows users from themselves, when in reality it's about protecting them from non-Windows users.

shikadilord commented:
Updated to account for `COM[5-9]` and `LPT[5-9]` (which a quick search appears to indicate are possible, at least on Windows NT/2000/XP and later), and added a comment to prevent people from accidentally breaking a check in the future.

jyrkive commented on src/addon/validation.cpp:
```diff
@@ -30,6 +30,16 @@ namespace {
+		"CONIN$", "CONOUT$",
+		// Configuration-dependent devices
+		"COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
+		"LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
```
There are so many numbered device names now that instead of hardcoding them, I'd prefer checking for DOS names with `(stem.find("COM") == 0 || stem.find("LPT") == 0) && stem[3] >= '0' && stem[3] <= '9'`

shikadilord replied:
The list isn't going to change any time soon and it's easier to tell what the items are this way.
